### DISCUSSION TITLE
NUMA-aware memory testing ?

### DISCUSSION THREAD

**Discussion by debrouxl** [memtest86plus/memtest86plus#12] [General]
It's been over a decade since multi-socket x86 computers with either AMD or Intel processors have become NUMA machines with 2+ NUMA nodes; nowadays, even mid-range and high-range single socket processors have 2+ NUMA nodes due to their respective chiplet-based designs.
Currently, pcmemtest / memtest86+ splices the address space without paying attention to memory/core affinity, effectively testing non-local access most of the time on servers with > 2 NUMA nodes equipped with symmetric amounts of memory. IIUC, some additional SMP tests attempting to perform local memory access on all NUMA nodes may increase the used memory bandwidth, and get closer to the limits of the controller(s) and the memory sticks on every node / socket.
Making NUMA-unaware and NUMA-aware modes mutually exclusive (assuming it makes sense to split them) and chosen at boot/startup time (cmdline or initial menu) might simplify the code, but it would force users to boot twice, which is less convenient.

AFAICT, the main source of appropriate information for memory/core affinity is ACPI System Resource Affinity Table ("SRAT" signature). System Locality Information Table ("SLIT" signature) is only an indication of how much slower non-local accesses are. When available, these tables can be extracted and decoded as follows by the ACPICA tools:

```
acpidump > acpidata.dat
acpixtract -sSRAT acpidata.dat
acpixtract -sSLIT acpidata.dat
iasl -d srat.dat
iasl -d slit.dat
```

`numactl -H` effectively outputs the same information as the SLIT, converted to decimal. For instance, on my 4S Opteron 62xx & 63xx servers with 3 HyperTransport links:
```
node   0   1   2   3   4   5   6   7 
  0:  10  16  16  22  16  22  16  22 
  1:  16  10  22  16  16  22  22  16 
  2:  16  22  10  16  16  16  16  16 
  3:  22  16  16  10  16  16  22  22 
  4:  16  16  16  16  10  16  16  22 
  5:  22  22  16  16  16  10  22  16 
  6:  16  22  16  22  16  22  10  16 
  7:  22  16  16  22  22  16  16  10 
```
4S Xeon setups with 2 QPI links (ring topology) probably have even more complex SLITs.

**Comment by x86fr** [comment]
Another important point to take into considerations is the new hybrid architecture with P-Core and E-Core. They are a nightmare to deal with because you have to heavily rely on Thread Director / HGS, both needing a complex software part that will not be available for Memtest86+. For ADL based CPU with E-Core, it's probably better to remove all E-Core from the thread list used for memory testing to avoid issues and slow performances.

**Comment by martinwhitaker** [comment]
pcmemtest/memtest86+ is testing the memory 1GB at a time. So all cores are concentrated on that 1GB. I don't see how you would get higher memory bandwidth utilisation by restricting it to just the local cores.

**Comment by debrouxl** [comment]
Higher bandwidth on the NUMA node to which the targeted 1 GB belongs, almost certainly not; however, higher total bandwidth and lower latency due to parallel test on multiple independent NUMA nodes, almost certainly, at least on multi-socket computers.
In the former situation, the upper bound for the memory bandwidth is the maximum bandwidth for a single socket and the latency for cache miss / writeback is memory + an inter-socket transfer, while in the latter situation, bandwidth maxes out at number of sockets * maximum bandwidth per socket, and latency for cache miss / writeback is memory latency.

**Comment by debrouxl** [comment]
There are noticeable speed changes during tests 1 and 2 (Address test, own address in window / Address test, own address + window) on my 4S PowerEdge R815 with 32 x 8 GB RAM sticks, and these changes are consistent with the ACPI SLIT+SRAT: for cores 0-7, the first 32 GB range is the fastest, and the last 32 GB range is among the slowest, as is the fourth; for cores 8-15, the second 32 GB range is the fastest one; etc.
The situation ought to be the same on the 4S RS904A with 16 x 16 GB sticks, which is enough to populate memory channels for all nodes, but I've never spent time examining the screen output for an extended period of time on these models.

IMO, cross-NUMA-node NUMA-unaware tests do make sense for most of the program's operation, but short of making all tests (optionally) NUMA-aware, a single new NUMA-aware test would be an interesting addition for minimizing memory wait, maximizing memory bandwidth, and therefore maximizing CPU temperature :)

**Comment by debrouxl** [comment]
Some data points about NUMA effects on my 4S RS904A with Opteron 6272 in 32 core mode (SMT disabled by BIOS), on the single test number 3:
* cores 0-31: 2 runs in 9'39"-9'40" each. Max CPU temperature 43-44°C;
* cores 0, 4, 8, 12, 16, 20, 24, 28 (one per NUMA node): 2 runs in 9'25"-9'26" each. Max CPU temperature 39-40°C;
* core 0, 0-33 GB (give or take, the memory associated to NUMA node 1): 1 run in 1'31";
* cores 0-1, 0-33 GB: 1 run in 0'55";
* cores 0-3, 0-33 GB: 1 run in 0'45".

I see three things here:
* 1st & 2nd figures: running with all cores is slower than running with 1 core per NUMA node, which adds insult to the injury of higher CPU temperature and higher power consumption;
* 5th figure: 8 NUMA nodes dealing with their own memory, one at a time, should therefore complete that test in around 8 x 45" ~6 minutes, i.e. less than two thirds of the run time of NUMA-unaware tests. 8 NUMA nodes dealing with their own memory in parallel should do significantly better, but the time can't be estimated accurately from the above figures;
* 3rd figure: a single core running `rep stosq` does not saturate the memory bus on that platform, as expected on most "modern" platforms.

**Comment by debrouxl** [comment]
Another data point on the usual RS904A in 32 core mode: Cores 0-3, 0-33 GB (NUMA node 1 and associated memory), all current tests (including the 1 and 2 own address tests) runs through the first pass in ~1h10', without nontemporal stores.
A NUMA-aware mode running through (enabled) NUMA nodes sequentially, one at a time, should therefore run in less than 10h on that computer; a parallel NUMA-aware mode should be able to do much better than that, but it would require an even larger rework of memtest86+... An additional first test where at least one core of every enabled NUMA node attempts to read and write the entire memory range (say, with mov_inv_fixed) to test reachability wouldn't take too long.
I'd expect a NUMA-aware mode to increase memory testing reliability, as I've already seen multiple computers with faulty RAM sticks or a memory access system below rating chug along for a while under idle conditions, and barf quickly after starting a higher-bandwidth workload.

**Comment by debrouxl** [comment]
I've been scratching that itch for the past 15+ days, as even halving the running time, without reducing test reliability AFAICT, on my 4S Opteron servers, would be an awesome result. Plus it would be a prerequisite for a parallel NUMA-aware mode, but IIUC, that would require one page directory per proximity domain, and implementing such a thing is a larger endeavour :)

From the data contained in the ACPI SRAT, I tried to build data structures appropriate for
1) splitting the VM map further to limit a range to a single proximity domain;
2) have only the cores which belong to that proximity domain work on the current VM map range (the others get an empty range and bail out immediately from the test functions);
3) hopefully leave the NUMA-unaware mode untouched for now, hence e.g. the current `chunk_index` / `chunk_index_in_proximity_domain` duplication, probably unnecessary in the end.

Maybe there are slightly different / better ways to achieve this aim.

I have published the current state of the work on the "numa" branch. I'm blocked from testing it on real hardware by a reproducible deadlock on the barriers, which I haven't been able to understand, by enabling barrier traces, using breakpoints (hence those `__asm__ volatile ("nop")` at select places...), or otherwise.
Usually, with a core and SRAT config mimicking those of my RS904A in BIOS-enforced no SMT mode, i.e. 8 proximity domains of 4 cores + 512 MB (instead of 32 GB...) of RAM each, I get the 4 first cores (belonging to proximity domain 0) stuck on the run barrier after entering `do_tick()`, and the 28 others stuck on the start barrier, though AFAICS, they do enter the test function.

The WIP QEMU+GDB debugging patch from #177 is very helpful for tracing through the code, though the addresses of variables in .data, .rodata and .bss aren't being relocated, which reduces usability a bit. I tried to fix that, by adjusting both the generated gdbscript to perform partial loads of the debugging data, and the ldscript to reorder sections, but I failed.

**Comment by debrouxl** [comment]
Slight progress tonight: I found that I can avoid the deadlock by commenting out calls to `do_tick()` and `flush_caches()` in `test_mov_inv_fixed()`.
This indicates that I need to find a way to obtain the right number of barrier entry and exits, or scrap the idea of building empty ranges in `calculate_chunk()`, which I think would be a shame, because it makes sequential NUMA support reasonably non-intrusive, mainly located in `setup_vm_map()` and `calculate_chunk()`.

**Comment by debrouxl** [comment]
There. The changes on the "numa" branch do provide significant speedups on both of my 4S Opteron Bulldozer & Piledriver server types, in both 32C/32T (RS904A) and 32C/64T (R815) mode :)
```
R815, 4 x Opteron 6378 (32C/64T), 32 x 8 GB DDR3-PC12800R = 256 GB, 70e49fc:
  numa, test 3: 01'08" -> 01'53" (33 GB, 00'45" = 045") -> 02'36" (65 GB, 00'43" = 043") -> 04'47" (161 GB, 02'11" = 131") -> 06'15" (225 GB, 01'28" = 088") -> 06'59" (total, 05'51" = 351")
  numa, test 4: 06'59" -> 10'27" (33 GB, 03'28" = 208") -> 13'54" (65 GB, 03'27" = 207")
  numa, test 5: 14'04" -> 17'41" (33 GB, 03'37" = 217")
nonuma, test 3: 00'22" -> 01'36" (33 GB, 01'14" = 074") -> 02'51" (65 GB, 01'15" = 075") -> 06'??" (161 GB, 03'??" = ???")->  08'44" (225 GB, ??'??" = ???") -> 09'58" (total, 09'36" = 576")
nonuma, test 4: 09'58" -> 15'54" (33 GB, 05'56" = 356") -> 21'49" (65 GB, 05'55" = 355")
nonuma, test 5: 22'19" -> 28'28" (33 GB, 06'09" = 369")
   
RS904A, 4 x Opteron 6272 (32C/32T), 16 x 16 GB DDR3-PC12800R = 256 GB, 70e49fc:
  numa, test 3: 00'45" -> 01'30" (33 GB, 00'45" = 045") -> 02'16" (65 GB, 00'46" = 046") -> 04'36" (161 GB, 02'20" = 140") -> 06'09" (225 GB, 01'33" = 093") -> 06'55" (total, 06'10" = 370")
  numa, test 4: 06'55" -> 10'34" (33 GB, 03'39" = 219")
  numa, test 5: 10'51" -> 14'49" (33 GB, 03'58" = 238")
nonuma, test 3: 01'38" -> 02'50" (33 GB, 01'12" = 072") -> 04'02" (65 GB, 01'12" = 072") -> 07'42" (161 GB, 03'40" = 220") -> 10'04" (225 GB, 02'22" = 142") -> 11'17" (total, 09'39" = 579")
nonuma, test 4: 11'17" -> 17'05" (33 GB, 05'48" = 348")
nonuma, test 5: 17'34" -> 23'33" (33 GB, 05'59" = 359")
```
-> nearly half faster on tests 3-5.

The NUMA-aware mode slows down the test slightly on a 2S Westmere server, though:
```
X8DTT-H, 2 x Xeon L5640 (12C/12T), 12 x 4 GB DDR3-PC10600R = 48 GB, 70e49fc:
  numa, test 3: 04'07" -> 04'33" (half, 00'26" = 026") -> 04'59" (total, 00'52" = 052")
  numa, test 4: 04'59" -> 07'04" (half, 02'05" = 125") -> 09'08" (total, 04'09" = 249")
  numa, test 5: 09'08" -> 11'18" (half, 02'10" = 130") -> 13'28" (total, 04'20" = 260")
nonuma, test 3: 04'07" -> 04'32" (half, 00'25" = 025") -> 04'57" (total, 00'50" = 050")
nonuma, test 4: 04'57" -> 06'49" (half, 01'52" = 112") -> 08'42" (total, 03'45" = 225")
nonuma, test 5: 08'42" -> 10'45" (half, 02'03" = 123") -> 12'49" (total, 04'07" = 247")
```

BTW, I did sanity-check the timings of the `nonuma` mode of the "numa" branch against 6.10, and they match:
```
X8DTT-H, 2 x Xeon L5640 (12C/12T), 12 x 4 GB DDR3-PC10600R = 48 GB, 6.10:
        test 3: 04'07" -> 04'32"                       -> 04'57" (total, 00'50" = 050")
        test 4: 04'57" -> 06'50"... 
        test 5: ...                                    -> 12'49"
```

Now to benchmark the other tests as well, a bit later.

**Comment by debrouxl** [comment]
I have also integrated the nontemporal accesses on test 2 (and therefore 1) into the numa branch. So far, the benchmarks of nontemporal + numa are looking good on the RS904A, as test 6 was reached ~24% / nearly an hour earlier than the baseline:
```
RS904A, 32C/32T, 860fcf5:  
         baseline    numa + nontemporal
test  0: 00:00:00              00:00:00
test  2: 00:00:28              00:00:29 (+00:00:01)
test  3: 02:07:47              01:49:51 (-00:17:56)
test  4: 02:17:27              01:56:10 (-00:21:17)
test  5: 03:03:51              02:25:32 (-00:38:19)
test  6: 03:52:10              02:56:07 (-00:56:03)
test  7: 10:03:30              07:01:02 (-03:02:28) - then deadlock and later manual computer reset, followed by selection of the sole 8-9-10 tests.
test  8: 10:47:36              00:00:00 (-10:47:36)
test  9: 11:34:22              00:36:38 (-10:57:44)
test 10: 15:04:01              02:54:02 (-12:09:59)
test  0: 15:14:13              03:04:16 (-12:09:57)
```
Half of test 6 was done in somewhere between 2h and 2h05, so it looks like test 7 should be reached around 7h run time / midnight CET by the numa + nontemporal run. EDIT: it was, indeed.
Test 7 deadlocks the computer, though, which means that the `do_tick()` occurrences are mismatched in that test. Will investigate and fix later.

EDIT: tests 8 and 9 work as well, and save another hour of run time. Since it's likely that a fixed NUMA-aware version of test 7 would be faster than the baseline, it looks like the NUMA-aware version reduces the pass time by more than 4 hours over a bit more than 15h - more than 25% faster.

**Comment by x86fr** [comment]
This looks very promising. 👌

This said, you really have to check the test efficiency (ability to detect x errors in y time), which is much more important than the time needed to run the test. 

The betatesting team has a process for this. It's not extremely reliable, but it works. Basically, you need to make your system unstable (usually by overclocking or undervolting the CPU or IMC/RAM), then compare the amount of errors detected in a fixed period of time between both builds). It's quite annoying because you have to find the right amount of "instability" (too high, the system will freeze or crash or reboot ; too low, it will take forever to grab enough data for a reliable result).

**Comment by debrouxl** [comment]
Yeah, for some classes of computers, this is promising.
The size penalty of the new data structures is currently reasonably low. 256 proximity domains is probably too high, but 16S/32S monster servers from HP, Atos (Bull) or other manufacturers, probably have at least 32 or even 64 proximity domains, and CXL memory might add further proximity domains.
For now, I'll focus on correctness first, ranging from not deadlocking the computer in the block move test by pairing `do_tick()` calls correctly, to ensuring that the code works properly on non-NUMA UP and SMP setups. However, you're right, the test you're mentioning needs to be done, even though the same bits are supposed to be written to memory (say, barring a silly mistake in my modifications to `calculate_chunk()`)... and I'm ill-equipped for that: in this thread, I've mentioned the three main NUMA computer models I have access to, two of them being basically twins, and I don't remember seeing overclocking / undervolting options in their respective BIOS menus.
Some of my desktop computers can do it, but they're single socket and proximity domain... and when I need to fiddle with their settings to improve their reliability, it's a matter of downclocking the RAM and/or CPU, without changing voltages :)

At this point, it's a matter of "NUMA-aware mode may make the test faster". I have a hunch that:
* a significant portion of 2S NUMA computer models will not benefit from it, as long as the higher memory bandwidth can mask the cross-proximity domain access penalty;
* 4S servers with Xeon E5-46xx v1-v4 (Sandy Bridge to Broadwell), the likes of HP DL560 G8-G9 or Dell PowerEdge R820-R830, may benefit, as these CPU models have only 2 QPI links (ring topology), whereas my 4S Opterons have 3 HT links (any socket to any). 4S IVB gear isn't that expensive, but I'm a bit reluctant to buy one (er, where would I be hosting it ? ^^).

4+S servers with high-end recent processors and high-bandwidth memory are currently inaccessible to memtest86+, due to the lack of x2APIC support. However, that's too much for me to tackle at the moment - and not only I don't have any such server model at home, but few of the computers I can reboot at will are x2APIC-capable anyway.

There are at least two other upsides to this NUMA support code, BTW:
* with the default `powersave=high` setting, it will usually make the test consume (much) less power, especially on (admittedly scarce) 4+S servers, which, even if the test runs for somewhat longer, may be a tradeoff acceptable to some users, in this day and age;
* the ACPI SRAT parsing code is one of the important building blocks for the NUMA node (proximity domain) selection capability I envisioned as #73 .

**Comment by debrouxl** [comment]
The NUMA-aware test is slightly faster on a common 2S server with fast 8C HSW and only 2+2 RAM sticks in the 12+12 RAM slots, which gives it sub-par bandwidth: 8 or 16 cores are significantly less than twice faster than 1 core (x2APIC mode, before I disabled it) !
And like all of the other computers on which I benchmarked them so far, nontemporal stores speed up test 2.

DL380 Gen9, 2 x Xeon E5-2667v3 (16C/16T, HT disabled at the BIOS level), 4 x 16 GB DDR4-PC17000R = 64 GB, 860fcf5:
```
         baseline, UP   numa + nontemporal, UP   baseline, SMP   numa + nontemporal, SMP
test  0: 00:00:00       00:00:00                 00:00:00        00:00:00
test  2: 00:00:02       00:00:02 (=00:00:00)     00:00:36        00:00:36 (=00:00:00)
test  3: 00:00:17       00:00:13 (-00:00:04)     00:04:49        00:03:45 (-00:01:04)
test  4: 00:01:39       00:01:35 (-00:00:04)     00:05:40        00:04:35 (-00:01:05)
test  5: --:--:--       00:08:16                 00:09:30        00:08:18 (-00:01:12)
test  6: --:--:--       00:15:04                 00:13:45        00:12:25 (-00:01:20)
test  7: --:--:--       01:07:55                 00:49:30        00:47:07 (-00:02:23) - then deadlock.
test  8: --:--:--       01:14:44                 00:52:40        00:00:00 (-00:52:40)
test  9: --:--:--       01:27:57                 00:57:28        00:04:33 (-00:52:55)
test 10: --:--:--       02:01:4? (cut by banner) 01:20:   (cut)  00:26:40 (>-00:53:20, probably ~00:53:31)
test  0: --:--:--       02:04:43                 01:23:25        00:29:54 (-00:53:31)
```

**Comment by debrouxl** [comment]
I bought a cheap used PowerEdge R820 with 4 x E5-4627v2 (summary of a nearly month-long long story which involved 2 servers - 1 returned - and 2 additional PSUs to replace the ones bricked by a non-cancellable firmware upgrade done by the replacement server's Lifecycle Controller upon its first two boots...), and I moved 16 x 16 GB of DDR3L-PC12800R into it, so as to populate all memory channels @ 1 DPC.

I know that 4S machines are uncommon, but still... NUMA effects are **spectacular** on this machine in a 4S configuration.
EDIT:
```
$ numactl -H # stripped output
available: 4 nodes (0-3)
node 0 cpus: 0 4 8 12 16 20 24 28
node 1 cpus: 1 5 9 13 17 21 25 29
node 2 cpus: 2 6 10 14 18 22 26 30
node 3 cpus: 3 7 11 15 19 23 27 31
node distances:
node   0   1   2   3 
  0:  10  20  30  20 
  1:  20  10  20  30 
  2:  30  20  10  20 
  3:  20  30  20  10 
```
Sure, I hoped that 2 QPI links (ring topology) would make a difference from 3 HT links (any-to-any topology), which is why I bought that model (R830 / DL560 G9 is another option, but much more expensive). However, I didn't expect a change of such a magnitude.

```
         numa + nontemporal, SMP, 4S   baseline, SMP, 4S
test  0: 00:00:00                      00:00:00            
test  2: 00:08:22                      00:08:19            
test  3: 02:01:00                     (00:24:04 skipped *1)
test  4: 02:03:21                     (00:51:47 skipped *2)
test  5: 02:14:12                     (00:58:47 skipped *3)
test  6: 02:25:47                     (--:--:-- skipped *4)
...
```
Explanations: 
1. test 2 reached 25% at 00:24:00, extrapolates to 100% reached around ~01:10:xx: nontemporal mode is a bit less than twice slower than normal mode. It's the first time I'm seeing nontemporal mode cause a slowdown in test 2 :'(
2. test 3 reached 50% at 00:45:54, extrapolates to 100% reached around 01:08:xx: **NUMA-aware mode > 15x faster than normal mode** !
3. the first GB of test 4 took 55" in normal mode, which extrapolates to ~4h - **NUMA-aware mode ~20x faster** !
4. the first GB of test 5 took 55" as well, so another ~4h extrapolated run time - likewise, **NUMA-aware mode ~<20x faster**.

I'll have to make tests in a 2S, 128 GB configuration, which is easy to obtain by removing the upper processor + RAM tray. But as it stands, NUMA-aware mode can test 256 GB of RAM in that server in much less than a day, while it would take closer to a couple days in normal mode.

Now if only I could make test 7 not deadlock the computer... and for good measure, I also need to fix the NUMA emulation for single proximity domain computers.

**Comment by debrouxl** [comment]
As expected, in 2S mode, NUMA effects aren't as strong on the R820:
```
R820, 16C/16T (2S mode), fbe2555, test 7 disabled:
         baseline    numa + nontemporal
test  0: 00:00:00              00:00:00
test  2: 00:01:27              00:01:27 (=00:00:00)
test  3: 00:12:22              00:19:14 (+00:06:52)
test  4: 00:13:51              00:20:22 (+00:06:31)
test  5: 00:20:45              00:25:18 (+00:04:33)
test  6: 00:28:06              00:30:49 (+00:02:43)
test  8: 01:25:26              01:17:26 (-00:08:00)
test  9: 01:32:38              01:23:31 (-00:09:07)
test 10: 02:12:48              01:53:17 (-00:19:31)
test  0: 02:17:11              01:57:41 (-00:19:30)
```
IOW, nontemporal writes still waste time on that platform, but NUMA awareness saves more than 26' over 137' after that, i.e. nearly 19% of run time.

**Comment by debrouxl** [comment]
@martinwhitaker when you have some time, I'll need help to move this forward :)
Test 7 keeps deadlocking, none of the ways I tried to avoid it worked. Also, the tick counting's progress is getting skewed: the progress bar still reaches 100% at the end of the test, but it progresses faster in the lower address ranges than in the higher address ranges.

**Comment by martinwhitaker** [comment]
Apologies, I don't much time to spare at the moment, but I will look at this when I can.

**Comment by debrouxl** [comment]
On the old K8 platform I've just mentioned in #320 , the code from this discussion also speeds up the test, by at least a minute over the ~15' reached before the computer locks itself out. Nontemporal writes decrease the time to reach test 3 from 43" to 30".

On the R820:
* in NUMA-aware mode, 256 GB of RAM (~0-258 GB range) are tested in ~7h50;
* in NUMA-aware mode, the 0-66 GB range (node 1, save a few dozen / hundred MBs) is tested in ~1h59;
* in normal mode, the 0-66 GB range was tested in ~13h53 !
* I haven't tested the full 0-258 GB range on the R820, since I'm using the computer for bursty workloads beyond its ~120W idle consumption, and ~56h at 600+W for something that can be done in less than 8h at significantly lower power is another matter in the spring and summer.

All values given with only temporal accesses, since non-temporal writes waste time on that platform.

**Reply by x86fr** [reply]
Interesting. Do you have a dispatch of gains by test for the first ~64GB?

**Reply by debrouxl** [reply]
Yup, I saved the timing data for the single benchmarking run I performed. Again, the debugging traces were partially cut off by the big PASS banner, which made me shorten "Test" to "T" to avoid the issue:
```
Dell Poweredge R820, 4 x Xeon E5-4627v2, 32 x 8 GB DDR3L-PC12800R (2.05 GB memory hole), 32 thread mode.
4 NUMA nodes of 64 GB, ring topology, SLIT indicates 10 / 20 / 30 / 20 and rotations thereof.
d9d87cd
Pass 1:
                         4 x 8C, normal          4 x 8C, NUMA               4 x 8C, NUMA
                         Range 0-66 GB (~node 1) Range 0-66 GB (~node 1)    Range 0-258 GB (nodes 1-4)
                         Temporal writes         Temporal writes            Temporal writes
 0 (test_addr_walk1)     00:00:00                00:00:00                   00:00:00
 1 (test_own_addr1)      --------                --------                   --------
 2 (test_own_addr2)      00:02:14                00:02:16 (+00:00:02) ~     00:08:31
 3 (test_mov_inv_fixed)  00:18:07                00:18:12 (+00:00:05) ~     01:11:25
 4 (test_mov_inv_fixed)  00:29:55                00:19:23 (-00:10:32) --    01:16:07
 5 (test_mov_inv_fixed)  01:28:49                00:24:37 (-01:04:12) --    01:37:01
 6 (test_mov_inv_walk1)  02:27:25                00:30:29 (-01:56:56) --    02:00:18
 7 (test_block_move)     --------                --------                   --------
 8 (test_mov_inv_random) 09:32:35                01:19:37 (-08:12:58) --    05:16:25
 9 (test_modulo_n)       10:27:48                01:25:56 (-09:01:52) --    05:41:40
10 (test_bit_fade)       13:50:?? (cut)          01:55:?? (-11:5?:??) --    07:42:?? (cut)
Pass 2:  
 0 (test_addr_walk1)     13:53:50                01:59:24 (-11:54:26) --    07:50:35
... (rebooted the server)


Core #0 / #4 / etc. prefer   0- 64 GB (memory hole not taken into account).
Core #1 / #5 / etc. prefer  64-128 GB.
Core #2 / #6 / etc. prefer 128-192 GB.
Core #3 / #7 / etc. prefer 192-256 GB.
```

NUMA awareness is supposed to have no impact on tests 0-2. 5" out of 18' is both within the benchmarking noise and higher variance than I usually see across memtest86+ runs.

**Comment by martinwhitaker** [comment]
@debrouxl , I've checked out your numa branch and run with qemu emulating 2 NUMA nodes, but haven't seen the hang in test 7. Do you have a qemu command line that reproduces it?

**Comment by debrouxl** [comment]
@martinwhitaker I can reproduce the deadlock in test 7 by using e.g. `MACHINE=2S12CWestmere MEMSIZE=... ./debug_memtest.sh` or `MACHINE=2S2CPenryn MEMSIZE=... ./debug_memtest.sh` from the `build64` directory, and in the EFI shell, passing the `numa` argument to memtest.efi:
```
fs0:
memtest.efi numa
```

With an added `set -x` before the final QEMU invocation, `MACHINE=2S2CPenryn MEMSIZE=6144 ./debug_memtest.sh` prints:
```
qemu-system-x86_64 -bios OVMF.fd -hda fat:rw:hda-contents -net none -drive if=pflash,format=raw,readonly=on,file=OVMF_CODE.fd -drive if=pflash,format=raw,file=OVMF_VARS.fd -machine q35 -m 6144M -cpu Penryn-v1 -smp 2,sockets=2,cores=1,maxcpus=2 -object memory-backend-ram,size=3072M,id=m0,prealloc=on -object memory-backend-ram,size=3072M,id=m1,prealloc=on -numa node,nodeid=0,memdev=m0,cpus=0 -numa node,nodeid=1,memdev=m1,cpus=1 -numa dist,src=0,dst=1,val=21 -numa dist,src=1,dst=0,val=21 -s -S
```

Screenshot with MACHINE=2S12CWestmere MEMSIZE=8192 and tests 2 + 7 enabled:
![image](https://github.com/memtest86plus/memtest86plus/assets/8375/95f8968b-23bb-47cb-8c53-fac8dd377b47)

My dream would be parallel testing of independent memory regions across NUMA nodes. Setting up page tables for that is the easy part of the work, though :)

**Reply by martinwhitaker** [reply]
Thanks. `debug_memtest.sh` doesn't work on my system, but I've reproduced the hang with the command line
````
qemu-system-x86_64 --enable-kvm -m 6G -smp 2 -object memory-backend-ram,size=3G,id=m0,prealloc=on -object memory-backend-ram,size=3G,id=m1,prealloc=on -numa node,nodeid=0,memdev=m0,cpus=0 -numa node,nodeid=1,memdev=m1,cpus=1 -numa dist,src=0,dst=1,val=21 -numa dist,src=1,dst=0,val=21 -cdrom memtest.iso -boot d
````
having made `enable_numa` default to true. Note that `--enable-kvm` makes qemu run a lot faster.

**Reply by debrouxl** [reply]
I do occasionally use `--enable-kvm` myself as well, for the same performance reasons, but I did not write it down here because it prevents the debugger integration part of `debug_memtest.sh` from working: the initial breakpoints are not being hit, etc.

**Reply by martinwhitaker** [reply]
Parallel testing will require a fundamental rethink of how memtest86+ performs tests. At the moment it relies on each thread doing the same thing and all threads proceeding in lockstep. If you allocate different regions of memory to different CPUs, there's no guarantee that each region is the same size - particularly with the various holes in the address map.

**Reply by martinwhitaker** [reply]
Test 7 is hanging because the middle part that moves the data around contains an inner loop
```` 
  for (int j = 0; j < iterations; j++) {
````
and the call to `do_tick()` is inside this loop, so is called `iterations` times. But CPUs that execute `SKIP_RANGE` only call `do_tick()` once.

**Reply by debrouxl** [reply]
Yeah, and I failed to fix that call count mismatch, despite trying several variants.

**Reply by martinwhitaker** [reply]
So I got it to work by
````
#define SKIP_RANGE(num_ticks) { if (my_cpu >= 0) { for (int i = 0; i < num_ticks; i++) { do_tick(my_cpu); BAILOUT; } } continue; }
 
````
then calling `SKIP_RANGE(iterations)` in that one place and `SKIP_RANGE(1)` everywhere else.

Not elegant, but it works.

**Reply by debrouxl** [reply]
It does, indeed - thanks :)

There remain two known bugs in this new version.
* in my understanding, the progress indicator remains skewed, and I think that it occurred after one of my deadlock fixes in a previous iteration, per the commit message. For instance, with MEMSIZE=5120 in 2S12CWestmere configuration:
    * on test 5, the 2 first GB of RAM (3 ranges: 0-4 MB, 4 MB-1 GB, 1 GB-2 GB minus a handful of pages reserved by memtest86+) represent 70% of the Test bar progress, and the 2 last GB of RAM (3 GB-sized ranges above 4 GB) represent the last 30% of the Test bar progress. I'd expect 50%/50%;
    * on test 5 or most other most tests, the last GB of RAM is reached when the Test bar percentage reaches 92% (slightly before 93%). With 6 memory ranges, I'd expect 83%.
* whatever the NUMA emulation code does wrong triggers a giant number of memory errors. I didn't spot the error in a cursory glance previously, this time I'll have to debug and fix it once and for all :)

**Reply by debrouxl** [reply]
* I think that the progress indicator skew is just a matter of the three first memory ranges being split into multiple pm chunks (0-A0000, etc.), whereas the latter memory ranges are not, so it doesn't look like a bug, in fact;
* I have fixed the NUMA emulation code for single proximity domain computers. I had forgotten a couple `<< PAGE_SHIFT`, since `pm_limit_lower` and `pm_limit_upper` are in page units.

Besides leftover debugging traces, one of which:
```
    for (int i = 0; i < vm_map_size; i++) {
        //do_trace(0, "vm %0*x - %0*x", 2*sizeof(uintptr_t), vm_map[i].start, 2*sizeof(uintptr_t), vm_map[i].end);
    }
```
helped me pinpoint both this remaining known bug and an issue in an intermediate version of my SIMD code (so it should probably stay for a little while), the NUMA awareness code looks in reasonable shape :)

**Reply by martinwhitaker** [reply]
Indeed, the progress indicator is not proportional to elapsed time. That's one of the minor irritations I'd like to fix one day.

**Comment by debrouxl** [comment]
I bought myself a cheap HPE DL380 Gen9. It's a departure from my 4S servers, but a more usual platform in the real world.
Unsurprisingly, for this server as well, in a 1 DIMM Per Channel configuration unlike the DL380 Gen9 I mentioned above, memtest86+ benefits from both NUMA awareness and nontemporal writes, again :)
```
HPE DL380 Gen9, 2 x Xeon E5-2667v4 @ 3.2 GHz, 8 x 16 GB DDR4-PC19200R (~ 2 GB memory hole), 16 thread mode (HT disabled, x2APIC disabled).
2 NUMA nodes of 64 GB, pair topology, SLIT indicates 10 / 21.
memtest86+ cd86295 (HEAD of simd branch as of 2023/11/19)
Pass 1:
                               (normal)     numa nontemporal          numa                 
test  0:                       00:00:00     00:00:00                  00:00:00             
test  1:                       00:01:10     00:01:10 (= 00:00:00)     00:01:10 (= 00:00:00)
test  2:                       00:10:00     00:07:05 (- 00:02:55)     00:10:01 (+ 00:00:01)
test  3:                       00:18:51     00:13:01 (- 00:05:50)     00:18:51 (= 00:00:00)
test  4:                       00:19:56     00:13:55 (- 00:06:01)     00:19:45 (- 00:00:11)
test  5:                       00:25:02     00:17:56 (- 00:07:04)     00:23:47 (- 00:01:15)
test  6:                       00:30:28     00:22:22 (- 00:08:06)     00:28:12 (- 00:02:16)
test  7:                       01:13:55     00:58:40 (- 00:15:15)     01:04:30 (- 00:09:25)
test  8:                       01:18:04     01:02:08 (- 00:15:56)     01:07:58 (- 00:10:06)
test  9:                       01:23:43     01:07:05 (- 00:16:38)     01:12:55 (- 00:10:48)
test 10:                       01:50:30     01:30:15 (- 00:20:15)     01:36:05 (- 00:14:25)
test 11:                       01:54:29     01:34:14 (- 00:20:15)     01:40:05 (- 00:14:24)
test 12:                       02:00:07     01:38:41 (- 00:21:26)     01:44:51 (- 00:15:16)
test 13:                       02:05:45     01:43:13 (- 00:22:32)     01:49:40 (- 00:16:05)
test 14:                       02:11:22     01:47:42 (- 00:23:40)     01:54:25 (- 00:16:57)
test  0:                       02:17:00     01:52:12 (- 00:24:48)     01:59:12 (- 00:17:48)
Average power consumption      ~260 W       ~245 W                    ~245 W               
Max peak power consumption     280 W        <270 W                    <270 W               
```
I'd have expected better power and thermal savings, but they exist nonetheless; the run is ~13% shorter with the sole NUMA awareness, and ~18% shorter with both NUMA awareness and nontemporal writes.
Nontemporal writes do not help in SISD mode (test 5 duration difference between numa and nontemporal is within measurement noise), but help in SIMD mode (differences between lines below test 10).

**Comment by debrouxl** [comment]
DL380p Gen8, 2 x Xeon E5-2670 0 @ 2.6 GHz, 16 x 8 GB DDR3-PC12800R (2 DPC, memory hole slightly larger than 1 GB), 16 thread mode (HT disabled).
Without NUMA awareness:
![memtest86+_DL380pG8_89d4b85_numa](https://github.com/memtest86plus/memtest86plus/assets/8375/59409565-248a-4b49-bf2a-cf3bd15051c0)
With NUMA awareness:
![memtest86+_DL380pG8_89d4b85_nonuma](https://github.com/memtest86plus/memtest86plus/assets/8375/a28ab109-b445-4b94-b36c-7c28639fdccc)
Nearly 24' saved over ~2h10' = 130'.
